In [1]:
#import libraries
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

#Read data then identify shape
df=pd.read_csv('C:\\Programming Projects\\Data Science Projects\\Project_1-Detecting_Fake_News\\news.csv')
df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [3]:
#DataFlair - Get labels
labels = df.label
labels.head()

#Split into training datasets and testing datasets
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)

## TfidfVectorizer

Convert a collection of raw documents to a matrix of TF-IDF features.

Equivalent to `CountVectorizer` followed by `TfidfTransformer`.

**input:** *{‘filename’, ‘file’, ‘content’}, default=’content’*

Learn more at https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html

In [4]:
#Initialize tfidVectorizer

#Stop words from the english language and max document frequency of 0.7.
#Stop words are the most common words to be filtered out before processing th natural language data.
tfidf_vectorizer = TfidfVectorizer(stop_words='english',max_df=0.7) 

#DataFlair - Fit and transform train set, transform test set
#fit_transform(raw_documents, y=None)[source]. Returns data frame.
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)


## PassiveAgressiveClassifier

Passive Aggressive algorithms are online learning algorithms. Such an algorithm remains passive for a correct classification outcome, and turns aggressive in the event of a miscalculation, updating and adjusting. Unlike most other algorithms, it does not converge. Its purpose is to make updates that correct the loss, causing very little change in the norm of the weight vector.

Learn more at https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.PassiveAggressiveClassifier.html

In [5]:
#DataFlair - Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)                  #max of 50 passes over training data
pac.fit(tfidf_train,y_train)                                  #fit the linear model with the PassiveAggressive alogirthm

#DataFlair - Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)                                # Predict class labels for samples in the testing data
score=accuracy_score(y_test,y_pred)                           # Return the mean accuracy on the given test data and labels.
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.98%


In [6]:
#DataFlair - Build confusion matrix
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[590,  48],
       [ 41, 588]], dtype=int64)

There are 590 true positives, 588 true negatives, 48 false positives, and 41 false negatives.